add helper.init() call to supply NR runtime path

mblackstock opened the pull request:
runtime initialization moved to a method so node tests can supply the node-red runtime path.
if it can find node-red in its dependencies, or it is running in the core, initialization is (still) not needed.

boneskull commented on README.md:
````diff
@@ -20,7 +20,7 @@ This will add the helper module to your `package.json` file as a development dep
 ```
 
-The test-helper requires the node-red runtime to run its tests, but Node-RED is **not** installed as a dependency.  The reason for this is that test-helper is (or will be) used in Node-RED core tests, and Node-RED itself has a large number of dependencies that you may not want to download if you already have it installed.
+The test-helper requires the node-red runtime, but Node-RED is **not** installed as a dependency.  The reason for this is that test-helper is (or will be) used in Node-RED core tests, and Node-RED itself has a large number of dependencies that you may not want to download if you already have it installed.
````
I'm not sure the intended average user of this module needs to the details about know why not.  That's documentation which could be better kept elsewhere, IMO.

boneskull replied:
Also the proper names here are inconsistent; we should probably mention `peerDependencies` or "peer dependency" right off the bat.

mblackstock replied:
Ok.  Will remove and add any useful info here to the 'creating a node' documentation at some point.

boneskull commented on README.md:
````diff
@@ -30,7 +30,7 @@ You can install the Node-RED runtime available for your unit tests one of two wa
 ```
 
-2. or link to Node-RED installed globally (recommended) using:
+2. or link to Node-RED installed globally using:
````
I've said this before, but I'm not sure we need to be in the business of suggesting this.  

I would prefer if those who wish to use `npm link` *help each other* with it, instead of adding a potentially flaky setup to the list of things we intend to support.

mblackstock replied:
OK will remove.

boneskull commented on README.md:
```diff
@@ -121,6 +123,10 @@ In this example, we require `should` for assertions, this helper module, as well
 
+## Initializing helper
+
+To get started, we need to tell the helper where to find the node-red runtime.  this is done by calling `helper.init(require.resolve('node-red'))` as shown.
```
capitalization 😜

boneskull replied:
I also wonder why we're doing this.  Can't we just `require('node-red')` then pass *that* to `helper.init()`?  Could that defeat the purpose of `helper.init()` entirely?

boneskull replied:
In other words: can we not reach all of the objects we need from the thing returned by `require('node-red')`?

boneskull commented on index.js:
```diff
@@ -22,28 +22,42 @@ var express = require("express");
-    var context = require(nrPath+"/red/runtime/nodes/context.js");
-    var events = require(nrPath+"/red/runtime/events.js");
+    // if not, we are running in the core
+    runtimePath = process.cwd()+"/red/red.js";
```
There are more robust ways to determine this.  For example, you could use something like [read-pkg-up](https://www.npmjs.com/package/read-pkg-up) then test if the `name` of the result is `node-red`.

boneskull replied:
That appending `.js` to a name in a `require()` call is not necessary, as it's assumed.  The only time this *is* necessary is if you have something like:

```
foo.js
foo/index.js
```

Then `require('foo')` does...what?  It's the same every time, but not obvious.

mblackstock replied:
Thanks - [read-pkg-up](https://www.npmjs.com/package/read-pkg-up) looks like a handy package.  Will see if I can rework things to use it.

Yes, I realize that, but the "main" of the node-red package is "red/red.js".  I need that module, as well as modules relative to "red" inside.  The path returned by `require.resolve('node-red')` is the full path to the red.js module (e.g. `/users/bob/my-node-package/node_modules/node-red/red/red.js` including the `red.js` at the end.  For consistency, I append `/red/red.js` to the current working directory if this fails, assuming I am running in the core.

Lines 52-57 get the path of the runtime if we are running in the core, or if node-red is a dependency.    This code does this synchronously when the helper is required so that existing core node tests need not call `init()`.

If I use [read-pkg-up](https://www.npmjs.com/package/read-pkg-up), I need to wait for results in a promise, so clients can't call helper methods until this is complete.  I will need to do this in the `load()` call I think...possibly others...

Anyway, I realize this is hacky, just need to take yet another step back...

mblackstock replied:
I see now that read-pkg-up can be called synchronously.  that makes things easier.

mblackstock commented:
@boneskull I think I've addressed all of your feedback.  Done some manual testing with using the helper in the node-red core on a branch, a contrib node, and supplied examples.  OK to merge?

boneskull commented on README.md:
```diff
@@ -6,55 +6,21 @@ Using the test-helper, your tests can start the Node-RED runtime, load a test fl
+    npm install node-red-node-test-helper node-red --save-dev
 
-This will add the helper module to your `package.json` file as a development dependency:
+This will add and the helper module to your `package.json` file:
```
typo

boneskull commented on package.json:
```diff
@@ -14,14 +14,15 @@
   },
   "peerDependencies": {
-    "node-red": "0.18.x"
+    "node-red": "~0.18.4"
```
can we be more forgiving than this?

mblackstock replied:
since we depend on the internals of the runtime currently, better safe than sorry I thought.
